libraries/BLE/src/BLECharacteristicImp.cpp

```diff
@@ -0,0 +1,749 @@
+    start->read = profile_read_process;
+
+    if (this->longCharacteristic() == false)
+    {
+        // Normal characteristic MAX. 20
+        start->write = profile_write_process;
+    }
+    else
+    {
+        // Long characteristic. MAX. 512
+        start->write = profile_longwrite_process;
+        start->flush = profile_longflush_process;
+    }
+    _attr_chrc_value = start;
+    pr_debug(LOG_MODULE_BLE, "chrcdescripor-%p, chimp-%p type-%d", start, this, this->type());
+    
+    start++;
+    index++;
+    counter++;
+    
+    if (this->subscribed())
+    {
+        // Descriptor
+        memset(start, 0, sizeof(bt_gatt_attr_t));
+        start->uuid = this->getClientCharacteristicConfigUuid();
+        start->perm = BT_GATT_PERM_READ | BT_GATT_PERM_WRITE;
+        start->read = bt_gatt_attr_read_ccc;
+        start->write = bt_gatt_attr_write_ccc;
+        start->user_data = this->getCccCfg();
+        
+        pr_info(LOG_MODULE_BLE, "cccd-%p", start);
+        
+        start++;
+        index++;
+        counter++;
+    }
+    
+    BLEDescriptorNodePtr node = _descriptors_header.next;
+    while (NULL != node)
+    {
+        BLEDescriptorImp *descriptorImp = node->value;
+        start = attr_start + index - base_index;
+        offset = descriptorImp->updateProfile(start, index);
+        counter += offset;
+        node = node->next;
+    }
+    pr_debug(LOG_MODULE_BLE, "%s:type-%d", __FUNCTION__, this->type());
+    return counter;
+}
+
+int BLECharacteristicImp::addDescriptor(BLEDescriptor& descriptor)
+{
+    BLEDescriptorImp* descriptorImp = descrptor(descriptor.uuid());
+    if (NULL != descriptorImp)
+    {
+        return BLE_STATUS_SUCCESS;
+    }
+    
+    descriptorImp = new BLEDescriptorImp(_ble_device, descriptor);
+    pr_debug(LOG_MODULE_BLE, "%s-%d",__FUNCTION__, __LINE__);
+    if (NULL == descriptorImp)
+    {
+        return BLE_STATUS_NO_MEMORY;
+    }
+    
+    pr_debug(LOG_MODULE_BLE, "%s-%d",__FUNCTION__, __LINE__);
+    BLEDescriptorNodePtr node = link_node_create(descriptorImp);
+    if (NULL == node)
+    {
+        delete[] descriptorImp;
+        return BLE_STATUS_NO_MEMORY;
+    }
+    link_node_insert_last(&_descriptors_header, node);
+    pr_debug(LOG_MODULE_BLE, "%s-%d",__FUNCTION__, __LINE__);
+    return BLE_STATUS_SUCCESS;
+}
+
+int BLECharacteristicImp::addDescriptor(const bt_uuid_t* uuid, 
+                                        uint16_t handle)
+{
+    BLEDescriptorImp* descriptorImp = descrptor(uuid);
+    if (NULL != descriptorImp)
+    {
+        return BLE_STATUS_SUCCESS;
+    }
+    
+    descriptorImp = new BLEDescriptorImp(uuid, handle, _ble_device);
+    pr_debug(LOG_MODULE_BLE, "%s-%d",__FUNCTION__, __LINE__);
+    if (NULL == descriptorImp)
+    {
+        return BLE_STATUS_NO_MEMORY;
+    }
+    
+    BLEDescriptorNodePtr node = link_node_create(descriptorImp);
+    if (NULL == node)
+    {
+        delete[] descriptorImp;
```

Comment:
Undefined behaviour. You cannot use `delete []` to free an object that was allocated with `new`. Use `delete`.